llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +152,304 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
+///   int case we track a value that represents the external Phi incoming
+///   value.
+/// * SmallPtrSet<Value*>: represent the constant (across updates) values
+///   that are used to update the pointer across different updates.
+/// * SmallPtrSet<Value*>: represent the chained counters across the updates.
+std::optional<IntrinsicChainInfo> trackIntrinsicChain(const Value *V) {
+  IntrinsicChainInfo ChainInfo;
+
+  if (!V->getType()->isPointerTy())
+    return std::nullopt;
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+    if (auto *GEP = dyn_cast<GEPOperator>(V)) {
+      if (GEP->hasAllZeroIndices()) {
+        // The result pointer and the first operand have the same value
+        V = GEP->getPointerOperand();
+      } else {
+        return std::nullopt; // No pointer arithmetic is allowed.
+      }
+    } else if (Operator::getOpcode(V) == Instruction::BitCast ||
+               Operator::getOpcode(V) == Instruction::AddrSpaceCast) {
+      V = cast<Operator>(V)->getOperand(0);
+      if (!V->getType()->isPointerTy())
+        return std::nullopt; // not a pointer.
+    } else {
+      if (auto *PHI = dyn_cast<PHINode>(V)) {
+        // Reached final destination.
+        return ChainInfo;
+      }
+      if (auto *AIEObject = lookThroughAIEAddressingModes(V)) {
+        IntrinsicArgsInfo NewArgSet = collectIntrinsicOperands(V);
+        // Initialize counters.
+        if (ChainInfo.IntrinsicParameters.CounterArgSet.empty()) {
+          ChainInfo.IntrinsicParameters.CounterArgSet = NewArgSet.CounterArgSet;
+        } else {
+          // if we already have counters, we need to reach the same
+          // AIEObject from the counters. Otherwise we can have some undesired
+          // counter update in between.
+          if (!isDirectReachable(ChainInfo.IntrinsicParameters.CounterArgSet,
+                                 AIEObject)) {
+            return std::nullopt; // not reachable.
+          }
+        }
+
+        // Update counters for the next round.
+        ChainInfo.IntrinsicParameters.CounterArgSet = NewArgSet.CounterArgSet;
+
+        // Initialize or compare other operands.
+        if (ChainInfo.IntrinsicParameters.ArgSet.empty()) {
+          ChainInfo.IntrinsicParameters.ArgSet = NewArgSet.ArgSet;
+        } else if (ChainInfo.IntrinsicParameters.ArgSet != NewArgSet.ArgSet) {
+          // Constant iterator information mismatch between sequence of
+          // intrinsic calls. We stop here.
+          return std::nullopt;
+        }
+        V = AIEObject;
+        ChainInfo.IntrinsicCalls++;
+      } else {
+        return std::nullopt; // something that we can't understand.
+      }
+    }
+    assert(V->getType()->isPointerTy() && "Unexpected operand type!");
+  }
+  return std::nullopt;
+}
+
+static AliasResult aliasAIEIntrinsic(const Value *ValueA, const Value *ValueB,
+                                     const Value *BaseA, const Value *BaseB) {
```

Comment:
I'm getting a bit confused with the different traversals we are doing: one for Value, one for Base, and then we also use `getUnderlyingObjectAIE` to compute `Base`. I would expect that here we do a more specialized traversal, so only `ValueA` and `ValueB` are needed.

I believe we could strart the traversal at both values using `trackIntrinsicChain()` and then compare the results:
 - Both traversals need to end on PHI nodes in the same BB
 - Both PHI nodes need the same initial pointer
 - The chain parameters must be the same: same number of calls, same initial counters, same parameters to the intrinsic calls

I think this is what you're doing here, but I'm wondering if it could not be siomplified into something like below in the end:
```
static AliasResult aliasAIEIntrinsic(const Value *ValueA, const Value *ValueB, unsigned LoopIterationDelta = 0) {
  std::optional<IntrinsicChainInfo> PATracking = trackIntrinsicChain(ValueA);
  std::optional<IntrinsicChainInfo> PBTracking = trackIntrinsicChain(ValueB);

  if (!PATracking || !PATracking)
    return MayAlias;

  return PATracking.samePointer(PBTracking, LoopIterationDelta) ? MustAlias : NoAlias;
}
```

Do you think something like this would be possible? Note we probably need to store a bit more in `IntrinsicChainInfo`,  like the base value we ended up at after the traversal in `trackIntrinsicChain`

Reply:
Hi @gbossu, indeed it looks a bit complicated. The base analysis is to prove the two pointers are updated in the same way (in a cyclic way). I will try a different approach that injects more information  in IntrinsicChainInfo and to not require 4 calls to trackIntrinsicChain.

Reply:
I think it looks great now :)